### PR TITLE
feat: 4018 feature add out of the box embedding support via clientfeedbackintegrations

### LINKED ISSUE #4018: [FEATURE] add out of the box `embedding` support via `client.feedback.integrations`
**Is your feature request related to a problem? Please describe.**
It is sometimes unclear what all the cool Argilla features are.

**Describe the solution you'd like**
Ideally, I want to see this happen when calling this `add_records` and run this code this before uploading data to Argilla. Use one of the [TaylorAI models](https://x.com/andersonbcdefg/status/1710708732534108413?s=20).
```python

rg.FeedbackDataset(..., vector_settings="TaylorAI/bge-micro-v2")
```

**Describe alternatives you've considered**
Doing this manually, always.

**Additional context**
https://github.com/argilla-io/argilla/issues/4017

### PR DESCRIPTION
**Pull request by davidberenstein1957**
# Description

I added support for `sentence-transformers` via the introduction of the `SentenceTransformersExtractor`. During the implementation, I fixed some minor issues.

- replaced the `vector_settings_by_name` that had been defined twice for the `remote` dataset and once within the `local` dataset, by the method we defined in the shared `base`.
- updated the `__repr__` of the dataset to include the `vector_settings`.
- rewrote some code of the `TextDescriptivesExtractor` to align the structure and usage of both extractors.
- resolved a bug in the `TextDescriptivesExtractor`, where we might update records from wrong indices with embedding during simultaneous IO.
- removed unit tests for the `TextDescriptivesExtractor` and created integration tests instead.
- removed some unused imoports
 
Closes #4018 

**Type of change**

(Please delete options that are not relevant. Remember to title the PR according to the type of change)

- [X] New feature (non-breaking change which adds functionality)
- [X] Refactor (change restructuring the codebase without changing functionality)
- [X] Improvement (change adding some improvement to an existing functionality)

**How Has This Been Tested**

(Please describe the tests that you ran to verify your changes. And ideally, reference `tests`)

- [X] integration/**/**/integration/test_textdescriptives.py
- [X] integration/**/**/integration/test_sentencetransformerx.py

**Checklist**

- [x] I added relevant documentation
- [x] I followed the style guidelines of this project
- [x] I did a self-review of my code
- [x] I made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK) (see text above)
- [x] I have added relevant notes to the `CHANGELOG.md` file (See https://keepachangelog.com/)

### CODE REVIEW

File: src/argilla/client/feedback/integrations/textdescriptives.py [hunk 1]
```diff
@@ -43,9 +44,8 @@ def __init__(
         self,
         model: str = "en",
         metrics: Optional[List[str]] = None,
-        fields: Optional[List[str]] = None,
-        visible_for_annotators: bool = True,
-        show_progress: bool = True,
+        visible_for_annotators: Optional[bool] = True,
```
Comment (davidberenstein1957) [inline src/argilla/client/feedback/integrations/textdescriptives.py:47]: @gabrielmbmb, do you think we should also move the `metrics` and `visible_for_anotators` to the `update_records` and `update_dataset` methods?

File: src/argilla/client/feedback/integrations/sentencetransformers.py [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+from argilla.utils.dependency import require_dependencies
+
+
+class SentenceTransformersExtractor:
+    """This class extracts a number of basic text descriptives from FeedbackDataset
+    records using the TextDescriptives library and adds them as record metadata."""
+
+    def __init__(
+        self,
+        model: Optional[Union["SentenceTransformer", str]] = "TaylorAI/bge-micro-v2",
```
Comment (gabrielmbmb) [inline src/argilla/client/feedback/integrations/sentencetransformers.py]:
```suggestion
        model: Union["SentenceTransformer", str] = "TaylorAI/bge-micro-v2",
```

This cannot be `None`, so no `Optional`

File: src/argilla/client/feedback/integrations/sentencetransformers.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+        self.show_progress = show_progress
+        if isinstance(model, str):
+            self.model = SentenceTransformer(model, **kwargs)
+        else:
+            if kwargs:
+                warnings.warn(
+                    "kwargs for initializing SentenceTransformer are ignored when passing a SentenceTransformer instance."
+                )
+            self.model = model
+        self.embedding_dim = self.model.encode(["sample sentence"])[0].shape[0]
```
Comment (gabrielmbmb) [inline src/argilla/client/feedback/integrations/sentencetransformers.py]:
```suggestion
        self.embedding_dim = self.model.get_sentence_embedding_dimension()
```

Comment (gabrielmbmb) [reply src/argilla/client/feedback/integrations/sentencetransformers.py]: https://github.com/UKPLab/sentence-transformers/blob/5c838a705c24c2dfd151a71674c99d09d014c1a9/sentence_transformers/SentenceTransformer.py#L375

File: src/argilla/client/feedback/integrations/sentencetransformers.py [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+
+    def _create_vector_settings(
+        self,
+        dataset: Union[FeedbackDataset, RemoteFeedbackDataset],
+        fields: Optional[List[str]],
```
Comment (gabrielmbmb) [inline src/argilla/client/feedback/integrations/sentencetransformers.py]:
```suggestion
        fields: List[str],
```

this is mandatory

File: src/argilla/client/feedback/integrations/sentencetransformers.py [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+        overwrite: bool = False,
+    ) -> Union[FeedbackDataset, RemoteFeedbackDataset]:
+        """
+        Create or update vector settings for the given dataset.
+
+        Args:
+            dataset (Union[FeedbackDataset, RemoteFeedbackDataset]): The dataset to create or update vector settings for.
+            fields (Optional[List[str]]): The list of fields to create or update vector settings for.
```
Comment (gabrielmbmb) [inline src/argilla/client/feedback/integrations/sentencetransformers.py]:
```suggestion
            fields (List[str]): The list of fields to create or update vector settings for.
```

File: tests/integration/client/feedback/integrations/test_sentencetransformers.py
Comment (gabrielmbmb) [inline tests/integration/client/feedback/integrations/test_sentencetransformers.py:1]: Need tests for checking the `raise ValueError(...)` conditions

File: src/argilla/client/feedback/dataset/base.py [hunk 6]
```diff
@@ -112,20 +109,31 @@ def question_by_name(self, name: str) -> Optional[Union[AllowedQuestionTypes]]:
...
+    def metadata_properties(
+        self,
+    ) -> Union[List["AllowedMetadataPropertyTypes"], List["AllowedRemoteMetadataPropertyTypes"]]:
         """Returns the metadata properties that will be indexed and could be used to filter the dataset."""
         pass
 
-    def metadata_property_by_name(self, name: str) -> Optional[Union["AllowedMetadataPropertyTypes"]]:
+    def metadata_property_by_name(
+        self, name: str
+    ) -> Optional[Union["AllowedMetadataPropertyTypes", "AllowedRemoteMetadataPropertyTypes"]]:
```
Comment (gabrielmbmb) [inline src/argilla/client/feedback/dataset/base.py]: Why making this type hint more complex? The remote metadata properties classes inherits from the non-remote ones so the type hint was correct.

Comment (gabrielmbmb) [reply src/argilla/client/feedback/dataset/base.py]: Same in the rest of methods

**Review by gabrielmbmb — CHANGES_REQUESTED** [review]
Hi @davidberenstein1957! Looks good. I wrote some comments about some type hints and missing unit tests.

**Pull request merged**